### PR TITLE
Fix detectron2 models on cpu-only torch

### LINKED ISSUE #1429: Detectron2 models loading issues with cpu-only torch builds
The current model factory of `detectron2` will fetch the default cfg settings, which uses cuda as cfg.MODEL.DEVICE, and convert to whatever `device` user specified while loading the models. One caveat is that `detectrons` models cannot be loaded properly if the underlying torch is cpu only. Therefore making the `detectron2` models unusable for torchbenchmark with cpu-only torch.

Repro:
```
# the benchmarks script is from pytorch repo, make sure to run this on cpu-only torch
python benchmarks/dynamo/torchbench.py --only detectron2_fasterrcnn_r_101_c4 -dcpu --performance 
```

The error message while loading `detectron2` models:
```
Torch not compiled with CUDA enabled
```

cc: @xuzhao9

### PR DESCRIPTION
**Pull request by shingjan**
This PR intends to fix #1429 .

cc:
@xuzhao9

### CODE REVIEW

**Comment by xuzhao9** [comment]
Thanks for the fix! Can you please add a test plan to this PR? For example:
 
```
python run.py detectron2_fasterrcnn_r_101_c4 -d cuda -t train
<output here...>
```

**Comment by shingjan** [comment]
@xuzhao9 Thanks for the prompt review! That line is removed in the PR and a test plan for this PR is as follows:
```
python run.py detectron2_fasterrcnn_r_101_c4 -d cpu -t eval

# output

Running eval method from detectron2_fasterrcnn_r_101_c4 on cpu in eager mode with input batch size 1.
/home/yj/pytorch/torch/functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at /home/yj/pytorch/aten/src/ATen/native/TensorShape.cpp:3483.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
CPU Total Wall Time: 3277.488 milliseconds
```

Note that I installed detectron2 from source instead of from pypi as there is a undefined symbol issues.

**Review by xuzhao9 — APPROVED** [review]
LGTM!

**Comment by xuzhao9** [comment]
@shingjan Can you take a look at the CI failure and fix it?

**Review by xuzhao9 — CHANGES_REQUESTED** [review]
Can you help fix the CI error?

File: torchbenchmark/util/framework/detectron2/model_factory.py
Comment (xuzhao9) [inline torchbenchmark/util/framework/detectron2/model_factory.py:94]: Should it be `cfg.MODEL.DEVICE = self.device`?
```suggestion
            cfg.MODEL.DEVICE = self.device
```

Comment (shingjan) [reply torchbenchmark/util/framework/detectron2/model_factory.py]: keen eyes! This is fixed

File: torchbenchmark/util/framework/detectron2/model_factory.py
Comment (xuzhao9) [inline torchbenchmark/util/framework/detectron2/model_factory.py:93]: What happens if cfg doesn't have the "MODEL" attr?

**Comment by shingjan** [comment]
@xuzhao9 really appreciate the review! Seems like workflow now needs maintainer's approval.

File: torchbenchmark/util/framework/detectron2/model_factory.py
Comment (shingjan) [reply torchbenchmark/util/framework/detectron2/model_factory.py:93]: From the [impl](https://github.com/facebookresearch/detectron2/blob/80e2673da161f57afe37ef769836a61976108ef1/detectron2/modeling/meta_arch/build.py#L23) of buliding detectron2 models, it looks to me that the assumption is that a `cfg` should always have the attribute `MODEL`. But per the last CI run, it seems that this assumption no longer holds at least for model `test_detectron2_fcos_r_50_fpn_eval_cuda`. Details [here](https://github.com/pytorch/benchmark/actions/runs/4258041173/jobs/7420088377#step:9:271). I am not familiar with this test and wonder what you recommendation would be for this case. Thanks!

**Comment by xuzhao9** [comment]
@shingjan Can you help fix the new CI errors?
You can test it locally with the command 

```
python test.py -k detectron2_fasterrcnn_r_101_c4
```

**Comment by shingjan** [comment]
@xuzhao9 With the latest change I ran `python run.py detectron2_fasterrcnn_r_101_c4 -d cpu -t eval` with output
```
Running eval method from detectron2_fasterrcnn_r_101_c4 on cpu in eager mode with input batch size 1.
/home/yj/pytorch/torch/functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at /home/yj/pytorch/aten/src/ATen/native/TensorShape.cpp:3488.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
CPU Total Wall Time: 3449.469 milliseconds
```
ran `python test.py -k detectron2_fasterrcnn_r_101_c4` with output
```
........
----------------------------------------------------------------------
Ran 8 tests in 48.829s

OK
```
Its still possible that CI has a different setup so the outputs may vary. Thx!

File: torchbenchmark/util/framework/detectron2/model_factory.py
Comment (xuzhao9) [reply torchbenchmark/util/framework/detectron2/model_factory.py:93]: Yes, this is because `detectron2_fcos_r_50_fpn_eval_cuda` doesn't use `MODEL.DEVICE` to initialize the model device. I am pretty sure this PR only enables the detectron2_fasterrcnn and detectron2_maskrcnn models for cpu-only torch, but [detectron2_fcos_r_50_fpn](https://github.com/pytorch/benchmark/tree/main/torchbenchmark/models/detectron2_fcos_r_50_fpn) is probably still GPU-only.

I am okay with that, and we can enable CPU with detectron2_fcos_r_50_fpn in another PR.

**Review by xuzhao9 — APPROVED** [review]
LGTM, thanks for the improvement.

**Comment by shingjan** [comment]
@xuzhao9 Thanks! Seems like CI for facebook internal failed, any hint that I can fix it on my end?

**Comment by xuzhao9** [comment]
@shingjan Those are a few timeout internal tests, no need to worry.

**Comment by shingjan** [comment]
@xuzhao9 sounds good! Is there a way we can merge this pr? Thx

**Comment by xuzhao9** [comment]
Sorry I have some rebasing issues with internal landing - can you help rebase this PR on the latest main branch and push again? Thanks! @shingjan

**Pull request closed without merging**
